cpp/leaky_core_error.cc

```diff
@@ -0,0 +1,72 @@
+// leaky_core_error.cc
```

Comment:
FYI sorry about the weird structure of this code, I noted it in the "smells" thread

https://oilshell.zulipchat.com/#narrow/stream/121539-oil-dev/topic/C.2B.2B.20Code.20Smells

This seems fine now but we will change the structure of errors a bit